libs/SalesforceSDK/src/com/salesforce/androidsdk/push/PushMessaging.kt

```diff
@@ -0,0 +1,466 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.push
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import com.google.firebase.FirebaseApp
+import com.google.firebase.FirebaseOptions
+import com.google.firebase.installations.FirebaseInstallations
+import com.google.firebase.messaging.FirebaseMessaging
+import com.salesforce.androidsdk.accounts.UserAccount
+import com.salesforce.androidsdk.app.SalesforceSDKManager
+import com.salesforce.androidsdk.util.SalesforceSDKLogger
+
+/**
+ * This class provides utility functions related to push notifications,
+ * such as methods for registration, unregistration, and storage and
+ * retrieval of push notification registration information from a
+ * private shared preference file.
+ */
+internal object PushMessaging {
+    private const val TAG = "PushMessaging"
+
+    // Public constants.
+    const val UNREGISTERED_ATTEMPT_COMPLETE_EVENT = "com.salesforce.mobilesdk.c2dm.UNREGISTERED"
+    const val UNREGISTERED_EVENT = "com.salesforce.mobilesdk.c2dm.ACTUAL_UNREGISTERED"
+    const val ACCOUNT_BUNDLE_KEY = "account_bundle"
+    const val ALL_ACCOUNTS_BUNDLE_VALUE = "all_accounts"
+
+    // Private constants.
+    private const val GCM_PREFS = "gcm_prefs"
+    private const val LAST_SFDC_REGISTRATION_TIME = "last_registration_change"
+    private const val REGISTRATION_ID = "c2dm_registration_id"
+    private const val BACKOFF = "backoff"
+    private const val DEVICE_ID = "deviceId"
+    private const val IN_PROGRESS = "inprogress"
+    private const val DEFAULT_BACKOFF: Long = 30000
+
+    /**
+     * Initiates push registration, if the application is not already registered.
+     * Otherwise, this method initiates registration/re-registration to the
+     * SFDC endpoint, in order to keep it alive, or to register a new account
+     * that just logged in.
+     *
+     * @param context Context.
+     * @param account User account.
+     */
+    @JvmStatic
+    fun register(context: Context, account: UserAccount?) {
+        val firebaseInitialized = initializeFirebaseIfNeeded(context)
+        if (firebaseInitialized) {
+            FirebaseMessaging.getInstance().isAutoInitEnabled = true
```

Comment:
FirebaseMessaging.getInstance() will throw an exception if called and the [DEFAULT] FirebaseApp has an invalid key or is null.